File: 1046-last-stone-weight/1046-last-stone-weight.java
class Solution {
    public static int ans(int[] stones){
        if(stones.length==1)return stones[0];
        PriorityQueue<Integer> pq = new PriorityQueue<>((a, b) -> b - a);
        for(int i=0;i<stones.length;i++){
            pq.add(stones[i]);
        }
        
        while(!pq.isEmpty()){
            int y = pq.peek();
            pq.remove(y);
            int x = pq.peek();
            pq.remove(x);
            pq.add(y-x);
            if(pq.size()==1) return pq.peek();
        }
        
        return 0;
    }
    public int lastStoneWeight(int[] stones) {
        int result = ans(stones);
        return result;
    }
}

File: 1046-last-stone-weight/1046-last-stone-weight.cpp
class Solution {
public:
    int lastStoneWeight(vector<int>& stones) {
        int n = stones.size();
        if(n==1){
            return stones[0];
        }
        
        priority_queue<int> pq;
        for(auto &it : stones){
            pq.push(it);
        }
        
        while(!pq.empty()){
            int y = pq.top();
            pq.pop();
            int x = pq.top();
            pq.pop();
            pq.push(y-x);
            if(pq.size()==1){
                return pq.top();
            }
        }
        
        return 0;
        
    }
};